src/compiler/rust.rs

```diff
@@ -0,0 +1,761 @@
+        }
+    } else {
+        None
+    };
+    // We'll figure out the source files and outputs later in
+    // `generate_hash_key` where we can run rustc.
+    CompilerArguments::Ok(RustHasher {
+        arguments: arguments.to_owned(),
+        filtered_arguments: filtered_arguments,
+        output_dir: output_dir.to_owned(),
+        externs: externs,
+        crate_name: crate_name.unwrap_or("<unknown>").to_string(),
+        dep_info: dep_info,
+    })
+}
+
+impl<T> CompilerHasher<T> for RustHasher
+    where T: CommandCreatorSync,
+{
+    fn generate_hash_key(self: Box<Self>,
+                         creator: &T,
+                         executable: &str,
+                         executable_digest: &str,
+                         cwd: &str,
+                         pool: &CpuPool)
+                         -> SFuture<HashResult<T>>
+    {
+        let me = *self;
+        let RustHasher { arguments, filtered_arguments, output_dir, externs, crate_name, dep_info } = me;
+        trace!("[{}]: generate_hash_key", crate_name);
+        let source_hashes = hash_source_files(creator, &crate_name, executable, &filtered_arguments, cwd, pool);
+        // Hash the contents of the externs listed on the commandline.
+        let cwp = Path::new(cwd);
+        trace!("[{}]: hashing {} externs", crate_name, externs.len());
+        let extern_hashes = hash_all(externs.iter()
+                                     .map(|e| cwp.join(e).to_string_lossy().into_owned())
+                                     .collect(),
+                                     &pool);
+        let creator = creator.clone();
+        let executable_digest = executable_digest.to_owned();
+        let executable = executable.to_owned();
+        let cwd = cwd.to_owned();
+        let hashes = source_hashes.join(extern_hashes);
+        Box::new(hashes.and_then(move |(source_hashes, extern_hashes)| -> SFuture<_> {
+            // If you change any of the inputs to the hash, you should change `CACHE_VERSION`.
+            let mut m = sha1::Sha1::new();
+            // Hash inputs:
+            // 1. A version
+            m.update(CACHE_VERSION);
+            // 2. The executable_digest
+            m.update(executable_digest.as_bytes());
+            // 3. The full commandline (self.arguments)
+            // TODO: there will be full paths here, it would be nice to
+            // normalize them so we can get cross-machine cache hits.
+            let args = arguments.iter().map(|s| s.as_str()).collect::<String>();
+            m.update(args.as_bytes());
+            // 4. The sha-1 digests of all source files (this includes src file from cmdline).
+            // 5. The sha-1 digests of all files listed on the commandline (self.externs)
+            for h in source_hashes.into_iter().chain(extern_hashes) {
+                m.update(h.as_bytes());
+            }
+            // 6. TODO: Environment variables:
+            //    RUSTFLAGS, maybe CARGO_PKG_*?
```

Comment:
> I wonder if it'd be possible to teach rustc to tell us about the usage of env!, similar to emitting dep-info

I was thinking the same thing yeah. I filed https://github.com/rust-lang/rust/issues/40364 to track this.

I also definitely agree that there's tons of junk in env vars typically. We could try to maintain a blacklist of "don't worry about caching this" but otherwise I think blanket hashing `CARGO_*` should work fine for now.